lib/Standalone/MapConv2DNhwcHwcfToMatmul.cpp

```diff
@@ -55,38 +56,89 @@ static bool hasFilterWithRandSEqualOne(OpOperand *filter, unsigned i,
   return ((filterShape[i] == 1) && (filterShape[j] == 1));
 }
 
+static FailureOr<Value>
+getSlicedConvOperandImpl(OpBuilder &builder, linalg::LinalgOp linalgOp,
+                         OpOperand *operand, ValueRange ivs,
+                         ValueRange valuesToUse, ArrayRef<int64_t> rAndSPos) {
+  Value operandToUse = valuesToUse[operand->getOperandNumber()];
+  ShapedType operandType = operandToUse.getType().cast<ShapedType>();
+  assert(operandType.hasStaticShape() && "tensor must have static shape");
+  size_t rank = operandType.getRank();
+  bool isImage = (operand->getOperandNumber() == 0) ? true : false;
+  unsigned desiredResultRank = 2;
+
+  SmallVector<OpFoldResult> offsets, sizes;
+  offsets.reserve(rank);
+  sizes.reserve(rank);
+
+  // offset into the tensor is the induction var or 0.
+  for (size_t idx = 0, e = ivs.size(); idx < e; idx++)
+    offsets.push_back(ivs[idx]);
+  for (size_t idx = ivs.size(), e = rank; idx < e; idx++)
+    offsets.push_back(builder.getIndexAttr(0));
+
+  OpOperand *filter = linalgOp.getInputOperands()[1];
+  if (isImage &&
+      !hasFilterWithRandSEqualOne(filter, rAndSPos[0], rAndSPos[1])) {
+    sizes = computeSizeGemmForImage(builder, linalgOp);
+  } else {
+    for (size_t idx = 0, e = rank - desiredResultRank; idx < e; idx++)
+      sizes.push_back(builder.getIndexAttr(1));
+    for (size_t idx = rank - desiredResultRank, e = rank; idx < e; idx++)
+      sizes.push_back(builder.getIndexAttr(operandType.getShape()[idx]));
+  }
+
+  // XXX: strides are assumed to be 1. This is not the case if
+  // the convolution has stride. Fix it.
+  SmallVector<OpFoldResult> strides(rank, builder.getIndexAttr(1));
+  return utils::getSliceOperand(builder, linalgOp, operandToUse, offsets, sizes,
+                                strides, desiredResultRank);
+}
+
+// Extract the sliced version of `operand` such that we can use it in a
+// linalg.matmul.
+static FailureOr<Value>
+getSlicedConvOperand(OpBuilder &builder, OpOperand *operand,
+                     linalg::LinalgOp linalgOp, ValueRange ivs,
+                     ValueRange valuesToUse, ArrayRef<int64_t> rAndSPos = {}) {
+  Location loc = linalgOp.getLoc();
+  FailureOr<SmallVector<Value>> involvedDimForOperand =
+      utils::getInvolvedLocalDimsForOperand(
+          builder, loc, operand, linalgOp.getMatchingIndexingMap(operand), ivs);
+  if (failed(involvedDimForOperand))
+    return failure();
+  return getSlicedConvOperandImpl(builder, linalgOp, operand,
```

Comment:
If you're not calling this from multiple sources, why is this not inlined?

Reply:
you mean `inline`? or you expect to be squashed with the caller?